### PR TITLE
[LinalgExt] Adding IndexingMaps to linalg_ext.attentionOp

### PR DESCRIPTION
**Pull request by raikonenfnu**
In order to make fusion with other generics, specifically transpose easier, we introduce affineMaps/indexingMaps to linalg_ext.attentionOp. With that we are also enforcing the number and types of dpsInputs. We are also removing "transpose_V" attribute in lieu of infering from indexingMaps.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 1]
```diff
@@ -1326,42 +1303,32 @@ LogicalResult AttentionOp::verify() {
...
   };
 
-  if (failed(checkShape("Query", getQueryType().getShape(), getQueryMap())) ||
-      failed(checkShape("Key", getKeyType().getShape(), getKeyMap())) ||
-      failed(checkShape("Value", getValueType().getShape(), getValueMap()))) {
+  if (failed(checkShape("Query", getQuery().getType().getShape(),
+                        getQueryMap())) ||
+      failed(checkShape("Key", getKey().getType().getShape(), getKeyMap())) ||
+      failed(checkShape("Value", getValue().getType().getShape(),
+                        getValueMap())) ||
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: The change looks unnecessary, and the `get*Type()` looks better to me. Why do you change them? Can you revert the change?

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 2]
```diff
@@ -1266,52 +1266,29 @@ LogicalResult WinogradOutputTransformOp::reifyResultShapes(
 //===----------------------------------------------------------------------===//
 
 LogicalResult AttentionOp::verify() {
-  Operation *op = getOperation();
+  AttentionOp attnOp = *this;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1269]: [No action item required] Technical, we don't need to capture `*this`. We can just do `emitOpError()` instead of (attnOp->emitOpError). I forgot why we have this style in the file. I'm just putting a note that it is not needed.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 3]
```diff
@@ -539,45 +541,64 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
...
-    Value getScale() {
-      return getDpsInputOperand(3)->get();
+    AffineMap getOutputMap() {
+      return cast<AffineMap>(getIndexingMapsArray()[3]);
+    }
+    int64_t getIterationDomainRank() {
+      return getQueryMap().getNumDims();
+    }
+
+    // TODO: Remove helper fns below once we move
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: dumb question: is fns == functions? If so, can you replace it with `functions` which is more common in the codebase?

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TileAttention.cpp [hunk 4]
```diff
@@ -145,6 +145,45 @@ static Value insertOutputSlice(Value src, Value dst,
...
+  AffineMap sumMap =
+      AffineMap::get(/*dimCount=*/4, /*symbolCount=*/0, {m}, ctx);
+
+  if (transposeV) {
+    SmallVector<AffineExpr> vDims(vMap.getResults());
+    std::swap(vDims[0], vDims[1]);
+    vMap = AffineMap::get(vMap.getNumDims(), vMap.getNumSymbols(), vDims, ctx);
+  }
+
+  SmallVector<AffineMap> attentionMaps{qMap, kMap, vMap, rMap, maxMap, sumMap};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TileAttention.cpp]: nit: Prefer using `()` constructor or `=` assignment instead of brace-list initialization. This kind of code is dangerous in some types. E.g.,

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.
```

I'd suggest to use either direct constructer or assignment style for consistency in the codebase.

See https://abseil.io/tips/88 for more details.

so it could be `SmallVector<AffineMap> attentionMaps = {qMap, kMap, vMap, rMap, maxMap, sumMap};`.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 5]
```diff
@@ -1385,52 +1352,8 @@ LogicalResult AttentionOp::reifyResultShapes(
...
-  bool isTiled = getNumResults() == 3;
-  if (isTiled) {
-    for (AffineMap &map : results) {
-      map = map.dropResult(0);
-    }
-  }
-
-  return results;
+  return SmallVector<AffineMap>(
+      getIndexingMaps().getAsValueRange<AffineMapAttr>());
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1354]: This is a new and good trick to me, thanks!

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 6]
```diff
@@ -641,40 +662,11 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
...
-    }
-    std::optional<AffineMap> getSumMap() {
-      if (getNumResults() < 3)
-        return std::nullopt;
-      return cast<AffineMap>(getIndexingMapsArray()[5]);
-    }
-
-    int64_t getIterationDomainRank() {
-      return getQueryMap().getNumDims();
+    bool getTransposeV() {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: Can you add some comments to it? I think people would be confused because you remove the whole `transpose_v` attribute and comments above.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 7]
```diff
@@ -539,45 +541,64 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
     FlashAttention 2 and optionally results in the current `max` and `sum`
     statistics while processing the current tile.
 
-    If transpose_v is speciifed, the V tensor passed as input is assumed to
-    be transposed:
-
     attention(Q, K, V, scale) = softmax(Q @ K.T * scale) @ V.T
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:549]: Can you help update the description? This paragraph is not completed to me. E.g., new people would ask what `V.T` is when they look at the description.

**Review by hanhanW — CHANGES_REQUESTED** [review]
Thanks for the patch! Just some nits, I reviewed all the files except TMTensor part. Because I have no ideas about TMTensor.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: SG will add it back. This was more for prototyping purpose as we change the way we initialize the DpsInputs i.e from SmallVector<ShapedType> to a bunch of shapedType. But no reason we can't have same helper functions. :)

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 7 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:549]: done! LMK if any of the notes need more details :)

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 6 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: done! lmk if any of the comments need more details. :)

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TileAttention.cpp (hunk 4 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TileAttention.cpp]: Thanks for the pro-tip!

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: Yeap! updated the comment, lmk if it's still too vague.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp]: done!

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 8]
```diff
@@ -541,6 +541,11 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_PureOp<"attention",
     FlashAttention 2 and optionally results in the current `max` and `sum`
     statistics while processing the current tile.
 
+    With indexing_maps we can also infer when variables are transposed.
+    for example, if V has indexing_map of (m, k1, k2, n) -> (n, k2) insted
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: typo: instead
```suggestion
    for example, if V has indexing_map of (m, k1, k2, n) -> (n, k2) instead
```

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 9]
```diff
@@ -642,39 +669,12 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
...
-    }
-
-    int64_t getIterationDomainRank() {
-      return getQueryMap().getNumDims();
+    // Use indexing_map to infer if V is transposed.
+    bool getTransposeV() {
+      AffineMap vMap = getValueMap();
+      auto vFastDim = llvm::cast<AffineDimExpr>(vMap.getResults().back());
+      int64_t nDim = vMap.getNumInputs() - 1;
+      return nDim != vFastDim.getPosition();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: If I read the comment correctly, are you looking for `isMinorIdentity` method? I.e., can we just do `return !vMap.isMinorIdentity();`?

>   /// Returns true if this affine map is a minor identity, i.e. an identity
  /// affine map (d0, ..., dn) -> (dp, ..., dn) on the most minor dimensions.
  bool isMinorIdentity() const;

https://github.com/llvm/llvm-project/blob/0171e23647d58f65c1cddbc22e16fda2bafa6e85/mlir/include/mlir/IR/AffineMap.h#L145-L147

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 6 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td]: How about rename the function to `isTransposeV`? `get*` is more like an accessor or returning a new object to me, while here we're checking if v is transposed or not.

Also, the comment could be updated to `Returns true if the indexing map of V is a minor transpose map`. What do you think?

**Review by hanhanW — APPROVED** [review]
LGTM, just few final nits. Thanks for addressing my comments.

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 9 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: Oh yeah, youre right! :)

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: @hanhanW actually turns out for minorIdentity would work in the base case of 
(m, k1, k2, n) -> (k2, n) since it needs to be contiguous identity dims, but for most part we'd have (b0, m, k1, k2, n) -> (b0, k2, n). Making it a non-minor identity.

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: I also tried projectedPermutation and minorIdentityWithBroadcast both which also doesn't cover some cases haha.

Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: oh I see, I forgot that there could be batch dimensions. Sorry for the bad suggestion and thanks for the updates in the comments.

**Review by hanhanW — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td (hunk 9 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:678]: haha nw! I was excited for that clean up too :D
